internal/monitors/statsd/statsd.go

```diff
@@ -0,0 +1,168 @@
+package statsd
+
+import (
+	"net"
+	"strconv"
+	"strings"
+)
+
+var metricSfPrefixMap = map[string]string{
+	"g":  "gauge.",
+	"c":  "derive.",
+	"ms": "latency.",
+	"s":  "gauge.",
+}
+
+type statsDListener struct {
+	ipAddr       string
+	port         uint16
+	tcp          bool
+	udpConn      *net.UDPConn
+	prefix       string
+	converters   []*converter
+	metricBuffer []string
+}
+
+type statsDMetric struct {
+	metricName string
+	metricType string
+	value      float64
+	dimensions map[string]string
+}
+
+func (sl *statsDListener) Listen() error {
+	if sl.tcp {
+		return sl.listenTCP()
+	}
+
+	return sl.listenUDP()
+}
+
+func (sl *statsDListener) listenUDP() error {
+	conn, err := net.ListenUDP("udp", &net.UDPAddr{
+		IP:   net.ParseIP(sl.ipAddr),
+		Port: int(sl.port),
+	})
+
+	if err != nil {
+		return err
+	}
+
+	sl.udpConn = conn
+	return nil
+}
+
+func (sl *statsDListener) listenTCP() error {
+	return nil
+}
+
+func (sl *statsDListener) FetchMetrics() []*statsDMetric {
+	rawMetrics := make([]string, len(sl.metricBuffer))
+
+	copy(rawMetrics, sl.metricBuffer)
+	sl.metricBuffer = nil
+
+	parsed := parseMetrics(rawMetrics, sl.converters, sl.prefix)
+
+	return parsed
+}
+
+func (sl *statsDListener) Read() {
+	chData := make(chan []byte)
+
+	if sl.tcp {
+		go sl.readTCP(chData)
+	} else {
+		go sl.readUDP(chData)
+	}
+
+	for data := range chData {
+		sl.metricBuffer = append(sl.metricBuffer, strings.Split(string(data), "\n")...)
+	}
+}
+
+func (sl *statsDListener) readTCP(chData chan []byte) {
+}
+
+func (sl *statsDListener) readUDP(chData chan []byte) {
+	// UDP needs to receive data packet by packet. Max packet size is 65535 for now.
+	buf := make([]byte, 65536)
+	for {
+		n, _, err := sl.udpConn.ReadFromUDP(buf)
+
+		if err != nil {
+			logger.WithError(err).Error("Failed reading UDP datagram.")
+			// Exit the loop if the connection is closed
+			if strings.Contains(err.Error(), "use of closed network connection") {
```

Comment:
This is just waiting to be a silent bug when we upgrade Go and they changed the error string and then we get goroutine leaks in the agent.

If you can't figure out a robust way to check the return value from `ReadFromUDP`, then set an atomic flag on the struct and check it to know if should shutdown.